components/eamxx/src/share/eamxx_model_init.cpp

```diff
@@ -0,0 +1,190 @@
+#include "eamxx_model_init.hpp"
+
+#include "share/io/scream_scorpio_interface.hpp"
+
+#include <regex>
+
+namespace scream
+{
+
+ModelInit::
+ModelInit (const std::vector<Field>& eamxx_inputs,
+           const std::shared_ptr<const AbstractGrid>& ic_grid,
+           const util::TimeStamp& t0)
+ : m_t0(t0)
+ , m_ic_grid (ic_grid)
+{
+  separate_inputs (eamxx_inputs);
+}
+
+void ModelInit::
+set_constant_fields (const std::vector<std::string>& fnames,
+                     const std::vector<double>&      fvalues,
+                     const std::shared_ptr<ekat::logger::LoggerBase>& logger)
+{
+  EKAT_REQUIRE_MSG (fnames.size()==fvalues.size(),
+      "Error! Inconsistent sizes for constant field initialization arrays.\n"
+      " - num fnames : " + std::to_string(fnames.size()) + "\n"
+      " - num fvalues: " + std::to_string(fvalues.size()) + "\n");
+
+  logger->info("    [EAMxx] Initializing constant fields ...");
+
+  // Helper lambda, to init all copies of a field (on all grids mgrs)
+  auto init_constant_field = [&] (const std::string& name, double val, int icmp = -1) {
+    Field* f = nullptr;
+    for (auto& it : m_ic_fields) {
+      if (it.name()==name) {
+        f = &it;
+        break;
+      }
+    }
+    EKAT_REQUIRE_MSG (f!=nullptr,
+        "Error! Attempt to prescribe constant value for a field that is not an atm input.\n"
+        " - field name: " + name + "\n");
+
+    if (icmp>=0) {
+      const auto& fl = f->get_header().get_identifier().get_layout();
+      EKAT_REQUIRE_MSG (fl.is_vector_layout(),
+          "Error! Attempt to access vector component of a non-vector field.\n"
+          "  - field name: " + name + "\n"
+          "  - field layout: " + to_string(fl) + "\n"
+          "  - requested comp: " + std::to_string(icmp) + "\n");
+      f->get_component(icmp).deep_copy(val);
+    } else {
+      f->deep_copy(val);
+    }
+    f->get_header().get_tracking().update_time_stamp(m_t0);
+  };
+
+  // Loop over provided fnames, and call the lambda above
+  for (size_t i=0; i<fnames.size(); ++i) {
+    logger->info("      " + fnames[i] + " = " + std::to_string(fvalues[i]));
+    auto pos = fnames[i].rfind("_cmp_");
+    if (pos==std::string::npos) {
+      // Not a component of a field
+      init_constant_field(fnames[i],fvalues[i]);
+    } else {
+      // Looks like the component of a vector field.
+      auto tail = fnames[i].substr(pos+5);
+      EKAT_REQUIRE_MSG (tail.size()>0,
+          "Error! Bad name specification for constant field initialization.\n"
+          " - input name: " + fnames[i] + "\n");
+      try {
+        int cmp = std::stoi (tail);
+        init_constant_field(fnames[i],fvalues[i],cmp);
+      } catch (std::exception&) {
+        EKAT_ERROR_MSG (
+          "Error! Bad name specification for constant field initialization.\n"
+          " - input name: " + fnames[i] + "\n");
+      }
```

Comment:
Is this only catching the case that `std::stoi(tail)` gives error? If so, should the error message mention this? I only say this because there is a high likelihood that a user reaches this error message if they are specifying constant fields. Something like 
```
"Error! Bad name specification for constant field initialization. "
"Cannot deduce field component from specified \"cmp_"+tail+"\"\n"
" - input name: " + fnames[i] + "\n"
```
The same for `EKAT_REQUIRE_MSG (tail.size()>0,...)` above.